File: problems-adapt/0100-0200/0110_fewest-square-summands/solution_dp.py
class Solution:
    def fewestSquareSummands(self, n: int) -> int:
        # dp[i] = fewest perfect squares summing to i: any decomposition ends
        # with some square s <= i, leaving the subproblem dp[i - s], so
        # dp[i] = 1 + min(dp[i - s]) over the squares precomputed here.
        squares = [i * i for i in range(1, int(n**0.5) + 1)]
        # dp[0] = 0 anchors the induction (zero squares sum to zero); the inf
        # sentinels simply lose every comparison until a real value lands.
        dp = [0] + [float("inf")] * n
        # Filling i in increasing order means every dp[i - s] consulted is
        # already final.
        for i in range(1, n + 1):
            for s in squares:
                if s > i:
                    break
                if dp[i - s] + 1 < dp[i]:
                    dp[i] = dp[i - s] + 1
        return dp[n]

File: problems-originals/0201-0300/0279_perfect-squares/solution_dp.cpp
class Solution {
  public:
    int numSquares(int n) {
        // dp[i] = fewest perfect squares summing to i: any decomposition ends
        // with some square s <= i, leaving the subproblem dp[i - s], so
        // dp[i] = 1 + min(dp[i - s]).
        vector<int> dp(n + 1, INT_MAX);
        // dp[0] = 0 anchors the induction (zero squares sum to zero).
        dp[0] = 0;
        // Filling i in increasing order means every dp[i - s] consulted is
        // already final; j*j <= i enumerates exactly the squares <= i.
        for (int i = 1; i <= n; ++i) {
            for (int j = 1; j * j <= i; ++j) {
                int s = j * j;
                if (dp[i - s] + 1 < dp[i])
                    dp[i] = dp[i - s] + 1;
            }
        }
        return dp[n];
    }
};
